### DISCUSSION TITLE
How to think about ReportPortal in a pull-request based workflow

### DISCUSSION THREAD

**Discussion by Bertg** [reportportal/reportportal#1626] [Q&A] [answered]
Hi,

This question might be a bit "vague" or totally missing the mark, but I have had issues finding an adequate answer.

So, we are running our current test suite on Github's workflow/action system. And we validate if pull request can be merged in based on the results of this runs. Now with this solution we don't get any insight into the evolution of our test suite (amount of tests, test time, etc...). We are looking for a solution for this, and we are considering ReportPortal for this. However I'm a bit at a loss of how we would map our current workflow to the project/launch model of ReportPortal.

Our test suit currently contains several tests, all of these get executed per pull request, and after merger on the main branch. Some of these tests are "Frontend Jest tests", "Backend test with spec", "Cypress for integration".

How should we model this? Is this a single project? Do we create a "launch" per PR/Branch?
Is there any best practice for this kind of setup? Or am I completely missing the ReportPortal use case?

Thanks for any help here!

**Comment by Bertg** [comment]
Just following up here. Does anyone have some insight into this?

**Comment by evjlobanova** [comment, marked as answer]
Hello @Bertg 

It really depends on what your needs.

Our recommendation regarding projects:
" The main question of how you can organize your working space in ReportPortal. Let's see two use cases which can help you to make a decision.

**Use Case 1:**

**Situation:** We have a small team. Everybody can see everything. We need one dashboard which shows information about the whole project.

**Solution:** Create one project. Assign all users on the created project. View all types of test cases (UI, API, Unit and so on) on this project. Divide executions by custom filters. Create a dashboard with a widget based on filters.

Pros: You have a dashboard that can combine all results needed for your team. You need to configure only one project.

Cons: All team members can see all results, you can not configure permissions rules.

**Use Case 2:**

**Situation:** We have a big team with a difficult permission structure and nested streams. Only persons with granted permissions can see the results of a certain stream.

Solution: Create separate projects for different streams. Assign users to appropriate projects. Create a dashboard with a widget based on filters for each project.

Pros: Security, only users with granted permission will have access to the data.

Cons: You can not see a single dashboard for several dashboards."
from https://reportportal.io/docs/Creation-of-project

Regarding how to organize launches and test suites:

The most common case:

1 launch equals to 1 test plan which is executed on a special env, platform - like "Frontend Jest tests on Chrome on Dev Env" and also you can specify different parameters using attributes type: UI, browser: Chrome, env: Dev

But launch naming depends on how you want to track [history](https://reportportal.io/docs/History-of-launches) and [widgets](https://reportportal.io/docs/Possible-Dasboards-in).

1. So if you will use launch name just as a Test Plan - so on the same history table and line I will see executions from different platforms and envs. If you do not have any platforms and envs. Just use Test Plan names or suites.

2. Majority of the widgets are based on the launches, if you want to use widgets you need to take into account this point.


So RP is a very flexible and will show you the info that you have sent.
I can try to send the first results and try to play with them.

**Comment by rdeangelis83** [comment]
Hi @evjlobanova,

do you have also an advice how to handle tests results from different branches?

We are executing (all) our tests on each PR in GitHub. 

Currently we are discussing whether it would be better to have the branch name as suffix in the launch name e.g. "MyBranch-RegressionAndIntegrationTests" or to add the branch as an attribute to the launch e.g "RegressionAndIntegrationTests" + attribute branch: "myBranch".

By having the branch name as part of the launch name it would be easier to see the history of test execution for one branch.
But on ther other hand some widgets like "Shaky Tests" will only check for one specific launch name.

Is there also other things that we should consider?

BR
Rocco
